Create Proto messages for SerializedDesignDoc

timothyfroehlich opened the pull request:
This is another step towards eliminating Serdegen. This should be the full set of messages required for SerializedDesignDoc.

The messages have been largely generated by Gemini, so there's likely some inaccuracies in there. Those will become evident when we start integrating the generated messages into the code, so I'm not too concerned about that.

What I would like review on right now is the layout of the files in the proto/ directory. Since we're trying to create a documented, stable format it made sense to split the messages into different packages. They are:

design: Everything needed for a DesignDoc
design/element: Basic elements of a design. Color, Rectangle, etc.
design/interaction: Animations and reactions
design/layout: Grid, alignment, etc.
design/modifier: Blends, filters, transforms
design/plugin: Figma Plugin specific, Dials/Guages data
toolkit: "Buisiness logic" types, currently SerializedDesignDoc.
android_interface: Messaging types that are only used to communicate across the JNI.

rylin8 commented on proto/design/layout_node.proto:
```diff
@@ -14,31 +14,12 @@
 
```
Not sure why this is not in the `layout` folder.

timothyfroehlich commented:
Ok, it's ready for review now. I've cleaned up the formatting, fixed some obviously bad messages, copied comments over and generated comments for the rest. I also noticed that instead of creating empty messages you can instead use the `google.protobuf.Empty` type, which seems to be a bit cleaner.

rylin8 commented:
The description still says that this is a draft.

timothyfroehlich commented:
> The description still says that this is a draft.

Fixed.

timothyfroehlich commented:
To be clear, the goal here is to get these messages checked in on their own, to keep the PR relatively small.

rylin8 commented on crates/layout/src/proto/layout_node.rs:
```diff
@@ -12,12 +12,17 @@
 // limitations under the License.
 
-use crate::layout_node::{self};
+use dc_proto::{
```
I'm confused because the crate `dc_proto` doesn't have much in it -- like it doesn't have any of the dependencies below in it. Can you explain somewhere how this all works?

rylin8 commented on proto/design/component.proto:
```diff
@@ -0,0 +1,47 @@
+message ComponentOverrides {
+  ViewStyle style = 1;
+  oneof component_content_override {
+    bool none = 2;
```
Are you able to use google.protobuf.Empty here instead of bool?

rylin8 commented on proto/design/element/variable.proto:
```diff
@@ -0,0 +1,94 @@
+message VariableValue {
+  oneof value {
+    bool boolean = 1;
+    float float = 2;
```
I'd call this "number"

rylin8 commented on proto/design/element/variable.proto:
```diff
@@ -0,0 +1,94 @@
+// symbolic name for a value that can change throughout your design.
+message VariableAlias {
+  string id = 1;
+  string name = 2;
```
I don't think a VariableAlias has a name in it. I think it's just the ID, and we have to find the name based off the ID. Though I'm not sure how you're using this, maybe you've already resolved the name.

rylin8 replied:
How was this resolved?

rylin8 replied:
`VariableAlias` is structured the way it is because that is how it is parsed from Figma. The 'type' corresponds to the string "VARIABLE_ALIAS", such as:
```
            "boundVariables": {
              "itemSpacing": {
                "type": "VARIABLE_ALIAS",
                "id": "VariableID:2039:72"
              },
            }
```

timothyfroehlich replied:
Wow, I'm sorry, Rob, I was not reading properly. I don't know why I looked at the VariableAlias message and thought that it had the same fields as the rust struct.

Ok, so you're saying that all we actually need is the ID. That makes sense. I also checked and BoundVariables doesn't get serialized (it's not in reflection.rs) so I shouldn't have a proto message for it anyways. I'll just remove the VariableAlias type since it's nothing more than a choice of the VariableValue enum/oneof.

rylin8 commented on proto/design/text_style.proto:
```diff
@@ -0,0 +1,49 @@
+  modifier.Background text_color = 1;
+  float font_size = 2;
+  optional string font_family = 3;
+  element.FontWeight font_weight = 4;
```
font size and weight can be variables now

rylin8 replied:
How was this resolved?

timothyfroehlich replied:
Ok. That's been fixed. I just set it to directly use a NumOrVar instead of a seperate message type, and we'll see if we want to change that when we work it into the code

timothyfroehlich reviewed (commented):
> How was this resolved? 

Sorry, I misinterpeted your comment as saying that they can be floats, but I see that you've merged in the font variables now.

timothyfroehlich commented on proto/design/element/font.proto:
```diff
@@ -0,0 +1,54 @@
+// The width of a font as an approximate fraction of the normal width.
+// Widths range from 0.5 to 2.0 inclusive, with 1.0 as the normal width.
+message FontStretch {
+  NumOrVar value = 1;
```
Clearly I wasn't thinking straight when I was addressing the last set of review comments.

rylin8 reviewed (approved):
Ok I think you've resolved all of my initial concerns, mainly with incorrect types in the proto files. I'm not sure I fully understand the organizational structure of where the proto files belong. For example `variable.proto` is in `element`, but `variable_collection.proto` is in `design`. There are a few others I'm not sure about, like the other proto fields directly in `design` but not in a subfolder. If you want to reorganize a bit later I'm ok with that too.

timothyfroehlich commented:
> Ok I think you've resolved all of my initial concerns, mainly with incorrect types in the proto files. I'm not sure I fully understand the organizational structure of where the proto files belong. For example `variable.proto` is in `element`, but `variable_collection.proto` is in `design`. There are a few others I'm not sure about, like the other proto fields directly in `design` but not in a subfolder. If you want to reorganize a bit later I'm ok with that too.

Yeah, I had the things in 'design' in a directory for a bit but couldn't figure out what to call it. I'll look at variable collection again. What else?

rylin8 commented:
Maybe put the stuff in `design` into `design/view`. Also, the fact that there are two `layout_style` protos is confusing.
